packages/ui-components/src/Button/Button.tsx

```diff
@@ -1,50 +1,70 @@
-import React, { FunctionComponent, ButtonHTMLAttributes } from "react";
+import React, { ElementType } from "react";
 import classnames from "classnames/bind";
 
 import objectToClassnames from "../utils/objectToClassnames";
 
 import styles from "./Button.module.scss";
 
 export type ButtonIntent =
-  | "primary-success"
+  | "primary"
   | "primary-danger"
   | "secondary"
   | "tertiary";
+
 export type ButtonSize = "standard" | "small";
 
-type OwnButtonProps = {
+type ElementProps<Type extends ElementType> = Pick<
+  React.ComponentProps<Type>,
+  keyof React.ComponentProps<Type>
+>;
+
+type CombinedProps<OwnProps, NativeElementProps> = OwnProps &
+  Omit<NativeElementProps, keyof OwnProps>;
+
+type OwnButtonProps<Element> = {
+  as: Element;
   intent?: ButtonIntent;
   size?: ButtonSize;
-  onClick?: () => void;
+  fluid?: boolean;
 };
 
-type NativeButtonProps = Omit<
-  ButtonHTMLAttributes<HTMLButtonElement>,
-  keyof OwnButtonProps
+type ButtonProps<OwnProps, Type extends ElementType> = CombinedProps<
```

Comment:
I think you can omit the double type parameter to `ButtonProps` and do:

```
type ButtonProps<Type extends ElementType> = CombinedProps<
  OwnButtonProps<Type>,
  ElementProps<Type>
>;
```

Because you want to make them always match.

In general, after playing around with this a bit, I don't think supporting an extra wide array of arbitrary HTML components is worth the type complexity.

I think we should just support `a` and `button` and extend from there. Then we can ensure a good experience with the component instead of making it infinitely customizable.

Reply:
I don't think we have that option (to only use `a` and `button`). We are internally linking with buttons all over the place.